pandas-stubs/core/frame.pyi

```diff
@@ -1832,9 +1836,9 @@ class DataFrame(NDFrame, OpsMixin):
     ) -> DataFrame: ...
     def pipe(
         self,
-        func: Callable[..., TType] | tuple[Callable[..., TType], _str],
-        *args,
-        **kwargs,
+        func: Callable[Concatenate[Self, P], TType] | tuple[Callable[..., TType], _str],
```

Comment:
I wish. I don't think Python's typing is flexible enough to account for the `tuple` version of the arg.

You'd want to express that _some_ argument (no longer just the first) in `func` takes in `Self`, and the rest of the params can be passed in via `args` and `kwargs`. This goes way beyond what `Concatenate` can do.

If you happen to know how to type this though happy to implement it. I'm just leaving it as is because I can't think of a way to improve it.

Reply:
So I took your PR, and used this:
```python
    func: Callable[Concatenate[Self, P], TType]
        | tuple[Callable[Concatenate[Self, P], TType], _str],
```
and it passed the typing tests.

If that hadn't worked, then I would have just using an `overload`, where you specify the 2 possible values of `func`, but I don't think that's necessary.

Reply:
Thanks for giving it a test drive!

But reading your annotation, this means that a `func` tuple argument can only take in a function as the first element if the function takes in `Self`. This is not explicitly true for the `tuple` version (it is for the `Callable` version of the arg though!).

The `tuple` version should take in a function as the first element where any one parameter (not necessarily the first) takes in `Self`.

I don't think we have a test right now that would explicitly used .

It's not obvious to me how `overload` would fix this either to be honest :(.

## Example of what should be acceptable behavior

```
import pandas as pd
from pandas import DataFrame

def hi(x: int, y: DataFrame) -> DataFrame:
    return y

pd.DataFrame({"a": [1, 2, 3]}).pipe((hi, "y"), 1)
```

Reply:
> The `tuple` version should take in a function as the first element where any one parameter (not necessarily the first) takes in `Self`.

You're correct.  So I think what you have is correct (with the `...`), but tests should be added.

Thanks for working on this!